Make verify_claims a separate function, added a post authentication signal

maerteijn opened the pull request:
Make it possible to reuse `verify_claims` when we, for example, want to use the refresh_token
to obtain a new access_token. Also send a `adfs_backend_post_authenticate` signal when authenticated for custom post authentication handling (like storing the tokens)

(BTW, excellent work with this package! Any chance you release the 1.0.x version on pypi soon?)

jobec commented:
Do you perhaps have an example of using a refresh token to fetch a new access token? Because as it is setup now, Django's session backend keeps you session alive, not the ADFS access token. Using the refresh token would mean you need to store it somewhere for each user, along with the lifetime of that token.

Same for the post authentication part. Some more detailed examples are welcome. All of that stuff might be interesting to implement directly in this package.

There are some cosmetic changes done also. Can you undo them? Just so there aren't unrelated things in the changeset.

As for version 1.0.x... I'm waiting for feedback on #54 about the DRF integration. Of that works sort of ok for others, I would like to take that along in v1.

maerteijn commented:
> Because as it is setup now, Django's session backend keeps you session alive, not the ADFS access token. Using the refresh token would mean you need to store it somewhere for each user, along with the lifetime of that token.

That is indeed correct.You could invalidate the django session (or in our case a custom auth header session) when the access token is expired and when it is not possible to retrieve a new one with the refresh token. In that case you would like to store both tokens in a (custom) model.

When you receive a new access token I want to validate the newly acquired access token again, that's why I would like to reuse the `validate_claims` method. (and it would be handy if this is a separate function, also if you want to write some more extensive tests for it).

> Do you perhaps have an example of using a refresh token to fetch a new access token? 

In our project we do it now like this:
```python
def refresh_access_token(refresh_token, request):
    data = {
        'grant_type': 'refresh_token',
        'client_id': settings.CLIENT_ID,
        'redirect_uri': provider_config.redirect_uri(request),
        'refresh_token': refresh_token,
    }

    # receive a new access code
    logger.debug("Receive a new access token with the refresh token")
    response = provider_config.session.post(
        provider_config.token_endpoint, data, timeout=provider_settings.TIMEOUT)

    if response.status_code != 200:
        raise PermissionDenied

    access_token = response.json()['access_token']
    logger.debug(f"Acquired new access token {access_token}")

    # and here I would like to revalidate the access token / claims
```
> 
> Same for the post authentication part. Some more detailed examples are welcome. All of that stuff might be interesting to implement directly in this package.

It could, but the refresh token -> new acces token part of this can be something really custom for a specific customer (specific models, or different flows), so I focussed now so I can easily extend / adjust functionality.

BTW, I changed the implementation a bit and added a signal instead of a separate function which has lesser impact (and still make it highly customizable for third party / customer specific projects)

However, I understand what you are saying so when I see something (later on) in our project that should be really in your package I would be happy to discuss / submit more pull requests of course.

>
> There are some cosmetic changes done also. Can you undo them? Just so there aren't unrelated things in the changeset.

I did undo them and keep your code as much intact as possible now
> 
> As for version 1.0.x... I'm waiting for feedback on #54 about the DRF integration. Of that works sort of ok for others, I would like to take that along in v1.

Ok, I would be happy with a release in the meantime of course but then I'll wait :)

jobec commented:
I agree it's better to put it in a separate function. But I don't like the concept of pulling it out of the authentication backend.
I rather think subclassing the django-auth-adfs backend and adding your own logic is a better thing to do.

So Sorry, I'll not be pulling this in as is. But I will work on separating the access token validation logic in a separate function.

maerteijn commented:
I now made `verify_claims` a member function of the backend.

maerteijn commented:
Ok, I now made it a member function of the authentication backend.

maerteijn commented:
👍 Thanks!
